### PR TITLE
cmd/cip: Default to non-production runs with `--confirm`

### PR DESCRIPTION
**Pull request by justaugustus**
#### What type of PR is this?

/kind bug regression
/priority critical-urgent

#### What this PR does / why we need it:

- test-e2e: Minor cleanup for cmp.Diff() output
- cmd/cip: Use `--confirm` instead of `--dry-run`

  Similar to our `--nomock` flags for other tools, we use a `--confirm`
  flag for `cip`, since the default nil value for booleans is `false`.
  
  Meaning: If `--dry-run` is not explicitly set to `true`, our tooling
  will initiate an image promotion.

- Flip confirm logic in cip/cip-auditor e2e MakeSyncContext() calls

/assign @hasheddan @xmudrii 
cc: @kubernetes-sigs/release-engineering 

#### Does this PR introduce a user-facing change?

```release-note
- cmd/cip: Use `--confirm` instead of `--dry-run`

  Similar to our `--nomock` flags for other tools, we use a `--confirm`
  flag for `cip`, since the default nil value for booleans is `false`.
  
  Meaning: If `--dry-run` is not explicitly set to `true`, our tooling
  will initiate an image promotion.
```

### CODE REVIEW

**Comment by justaugustus** [comment]
(Needed to flip the flag in the e2e test as well. Fixed in https://github.com/kubernetes-sigs/k8s-container-image-promoter/pull/285/commits/6a5b04ae3e69f2bfaed0d2737547d49cdf8f2ed8.)

**Comment by justaugustus** [comment]
```
 checking snapshots BEFORE promotion:
execing cmd bazel [run --workspace_status_command=/home/prow/go/src/sigs.k8s.io/k8s-container-image-promoter/workspace_status.sh //cmd/cip:cip -- run --snapshot=us.gcr.io/k8s-cip-test-prod/golden-bar]
time="2020-12-01T00:58:09Z" level=fatal msg="<<<<<<< got (type []inventory.Image)\n[{bar map[sha256:610b1ef6fec876146dee2b2846c890b566d26f235d7ea8982056a3e84bd35929:[1.0]]}]\n=======\n[]\n>>>>>>> expected (type []inventory.Image)" 
```

???

*grumbles*
/test pull-cip-e2e

**Comment by justaugustus** [comment]
Local run:

```console
$ time bazel run --workspace_status_command=$(pwd)/workspace_status.sh //cmd/cip:cip -- run --snapshot=us.gcr.io/k8s-cip-test-prod/golden-bar
INFO: Analyzed target //cmd/cip:cip (0 packages loaded, 0 targets configured).
INFO: Found 1 target...
Target //cmd/cip:cip up-to-date:
  bazel-bin/cmd/cip/cip_/cip
INFO: Elapsed time: 2.650s, Critical Path: 2.49s
INFO: 13 processes: 13 linux-sandbox.
INFO: Build completed successfully, 14 total actions
INFO: Build completed successfully, 14 total actions
INFO Request {us.gcr.io/k8s-cip-test-prod/golden-bar   true}: OK 
INFO Request {us.gcr.io/k8s-cip-test-prod/golden-bar/bar   false}: OK 
- name: bar
  dmap:
    "sha256:610b1ef6fec876146dee2b2846c890b566d26f235d7ea8982056a3e84bd35929": ["1.0"]

real	0m4.111s
user	0m0.125s
sys	0m0.064s
```

**Comment by justaugustus** [comment]
Oh hello, auditor flake?
/test pull-cip-auditor-e2e

**Review by hasheddan — APPROVED** [review]
/lgtm

**Review by saschagrunert — APPROVED** [review]

**Comment by saschagrunert** [comment]
The CI issue seems reproducible to me. :thinking:

**Comment by justaugustus** [comment]
Superseded by https://github.com/kubernetes-sigs/k8s-container-image-promoter/pull/291.
/close

**Comment by justaugustus** [comment]
Reset this branch to be built on top of https://github.com/kubernetes-sigs/k8s-container-image-promoter/pull/291.
/hold until that merges and that content is rebased out.

**Comment by justaugustus** [comment]
Bazel fetch flake for rules_python:
/test pull-cip-auditor-e2e

**Comment by listx** [comment]
Looking at this again, can we keep the old `--dry-run` flag? Why can't we additionally add a `--no-dry-run` flag that simply aliases into the old `--dry-run` flag (that is, make `--no-dry-run` be the same thing as `--dry-run=false`)?

>  Meaning: If `--dry-run` is not explicitly set to `true`, our tooling will initiate an image promotion.

If we're by default trying to make the tool write (promote) into GCRs, I am against this approach. The default invocation (without setting any special flags) should be a dry run; this been the case for this tool for a long time (at least for the majority of the time that this tool has been in use) and I don't want to change semantics for folks who might not be as familiar with the tool.

In addition, I personally do not like the whole `--nomock` negative boolean semantics, and would not look to spread what I think was originally a design decision that was created in anago (the shell script). (Even if we want to follow `--nomock`-based semantics --- isn't that the same thing as what we currently do today, requiring users to provide `--dry-run=false` to actually perform a promotion rather than just dry-running by default?)

@justaugustus @saschagrunert @spiffxp  WDYT?

/hold

**Comment by spiffxp** [comment]
We don't have a clearly consistent pattern across all of the tools used in this project.  While I don't like the double-negative of "no dry run false" I personally am a fan of what the flag is trying to accomplish:
- defaulting to a dry-run mode of operation for tools that mutate the state of the world via an API
- lack of explicit initialization implying false (this is how golang works for booleans, and feels closer to how bash would handle `[[ -n ${foo} ]]` or `${foo} == "true"`)

We've used `--confirm` for these semantics in a number of tools:
- https://github.com/kubernetes/k8s.io/blob/main/groups/reconcile.go#L131 - groups reconciler
- https://github.com/kubernetes/test-infra/blob/master/label_sync/main.go#L138 - label_sync
- https://github.com/kubernetes/test-infra/blob/master/prow/cmd/branchprotector/protect.go#L92 - branchprotector
- https://github.com/kubernetes/test-infra/blob/master/prow/cmd/peribolos/main.go#L88 - peribolos
- https://github.com/kubernetes/test-infra/blob/master/robots/commenter/main.go#L69 - commenter
- https://github.com/kubernetes/test-infra/blob/master/robots/pr-creator/main.go#L86 - pr-creator

There are, of course, just as many prow components that use `dry-run` and default it to `true` (hook, horologium, prow-controller-manager, sub, tide) because the confirm convention came about after a number of components had been written, and isn't consistently enforced.

The `--dry-run` and `--no-dry-run` convention feels more pythonic, but would be less easy to consistently implement across languages, and I'm not finding prior art for it in other tools on this project atm.

cs.k8s.io links for reference:
- [confirm](https://cs.k8s.io/?q=%22confirm%22&i=nope&files=&excludeFiles=&repos=)
- [dry-run](https://cs.k8s.io/?q=%22dry-run%22&i=nope&files=&excludeFiles=&repos=)
- [no-dry-run](https://cs.k8s.io/?q=%22no-dry-run%22&i=nope&files=&excludeFiles=&repos=)

tl;dr I would prefer `confirm` over `no-dry-run`, or reverting back to `dry-run (default: true)`

**Comment by justaugustus** [comment]
> We don't have a clearly consistent pattern across all of the tools used in this project. While I don't like the double-negative of "no dry run false" I personally am a fan of what the flag is trying to accomplish:
> 
>     * defaulting to a dry-run mode of operation for tools that mutate the state of the world via an API
> 
>     * lack of explicit initialization implying false (this is how golang works for booleans, and feels closer to how bash would handle `[[ -n ${foo} ]]` or `${foo} == "true"`)

@spiffxp -- You nailed it here! The goal is ensure the tool defaults to dry-run and requires additional work to trigger a production promotion.

> tl;dr I would prefer `confirm` over `no-dry-run`, or reverting back to `dry-run (default: true)`

`--confirm` sounds good to me; I'll update the PR to that effect.

**Comment by listx** [comment]
> > We don't have a clearly consistent pattern across all of the tools used in this project. While I don't like the double-negative of "no dry run false" I personally am a fan of what the flag is trying to accomplish:
> > ```
> > * defaulting to a dry-run mode of operation for tools that mutate the state of the world via an API
> > 
> > * lack of explicit initialization implying false (this is how golang works for booleans, and feels closer to how bash would handle `[[ -n ${foo} ]]` or `${foo} == "true"`)
> > ```
> 
> @spiffxp -- You nailed it here! The goal is ensure the tool defaults to dry-run and requires additional work to trigger a production promotion.

To be clear, the tool already defaults to `--dry-run=true`, even if you don't specify the flag.

**Comment by spiffxp** [comment]
To @listx's point, I think this change on its own can come across as churn.  If there is a broader effort to standardize on this pattern across tools, then it seems like there's more value in doing this.

**Comment by spiffxp** [comment]
I feel like maybe it _used_ default to true and then somewhere in the churn of https://github.com/kubernetes-sigs/k8s-container-image-promoter/pull/273 that got lost?

**Comment by justaugustus** [comment]
Slack convo: https://kubernetes.slack.com/archives/CJH2GBF7Y/p1629138583178000

**Comment by listx** [comment]
> I feel like maybe it _used_ default to true and then somewhere in the churn of #273 that got lost?

Yeah that seems likely. FTR every deployed instance of CIP in recent memory (including the current version v2.4.1 that's seeing presubmit/postsubmit usage) has always had `--dry-run=true` as the default.

**Comment by justaugustus** [comment]
> Yeah that seems likely. FTR every deployed instance of CIP in recent memory (including the current version v2.4.1 that's seeing presubmit/postsubmit usage) has always had `--dry-run=true` as the default.

Right, we haven't changed the actual `cip run` invocations for CI jobs, but the default behavior/underlying implementation is what needs to be fixed.

**Comment by justaugustus** [comment]
Just to capture some things from Slack:

This line creates a new set of options:
https://github.com/kubernetes-sigs/k8s-container-image-promoter/blob/97e351eb08d2af791f1a62ead66e5291398c4f0b/cmd/cip/cmd/root.go#L38

The options themselves: https://github.com/kubernetes-sigs/k8s-container-image-promoter/blob/97e351eb08d2af791f1a62ead66e5291398c4f0b/legacy/cli/root.go#L25-L28

If `--dry-run` is never set, it will be its `nil` value, which is `false` AKA unconditional promotion, which is what we don't want.
We'll proceed w/ `--confirm` as the flag name (to improve readability and make it a little more consistent with other tools across the community).

**Comment by justaugustus** [comment]
(rebasing for handoff)
I've asked @puerco to take over this branch/work in https://kubernetes.slack.com/archives/CJH2GBF7Y/p1631539082071700.
/assign @puerco

**Comment by justaugustus** [comment]
@puerco -- Nice work! 🎉 
/hold cancel

**Comment by puerco** [comment]
Reviewing @justaugustus commits :)
/lgtm

**Comment by justaugustus** [comment]
Symbolic LGTM for @puerco's fixes (even though the bot will yell at me):
/lgtm
/approve
/honk
